src/coreclr/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs

```diff
@@ -126,7 +128,53 @@ private void StartCallback()
         [MethodImpl(MethodImplOptions.InternalCall)]
         private static extern void SpinWaitInternal(int iterations);
 
-        public static void SpinWait(int iterations) => SpinWaitInternal(iterations);
+        public static void SpinWait(int iterations)
+        {
+            if (iterations <= 0)
+            {
+                return;
+            }
+
+            if (!ArmBase.IsSupported && !X86Base.IsSupported)
+            {
+                SpinWaitInternal(iterations);
+                return;
+            }
+
+            // The algorithm below is ported from native's
+            // void YieldProcessorNormalized(const YieldProcessorNormalizationInfo &normalizationInfo, unsigned int count)
+
+            // Prevent overflow on the multiply below on 32bit
+#if TARGET_32BIT
+            // Copied from YieldProcessorNormalization (yieldprocessornormalized.h)
+            const int targetNsPerNormalizedYield = 37;
+            const int maxYieldsPerNormalizedYield = targetNsPerNormalizedYield * 10;
+            const int maxCount = int.MaxValue / maxYieldsPerNormalizedYield;
+            if (iterations > maxCount)
+            {
+                iterations = maxCount;
+            }
+#endif
+
+            nint n = (nint)iterations * GetYieldsPerNormalizedYield;
+            Debug.Assert(n != 0);
+            do
+            {
+                if (ArmBase.IsSupported)
+                {
+                    // TODO: on some CPUs it's better to call ISB here
+                    ArmBase.Yield();
+                }
+                else if (X86Base.IsSupported)
+                {
+                    X86Base.Pause();
+                }
+                else
+                {
+                    throw new PlatformNotSupportedException();
```

Comment:
Will this normally be elided by the jit?

Reply:
Yep, this was mostly needed to satisfy our SIMD analyzer that got confused in this method